### PR TITLE
Integrate Transaction Model into the Library

### PR DESCRIPTION
**Pull request by bfish713**
- Consolidate: `TransactionHistoryEntry`, `TransactionInfo`,` PendingTransaction`, and the associated indices into just using the Transactions Model
- Remove `TransactionDatabase`
- Use `TransactionParams` for building the transaction before submission 

There also had to be some logic changes to use the new model.  The most notable are:
- We need to check the transaction type before adding freeze/unfreeze records to our internal records db
- When we removed expired transactions we need to check the transaction's status to make sure it's still pending
- For transaction resubmission we can skip storing the transaction and setting the hash since it will already be in our database.  This is done by optionally passing TransactionParams to `submit_elaborated_txn`.  We'll pass the params if it's a new transaction and if not (resubmission) we'll just load the transaction from storage.

#Closes: https://github.com/EspressoSystems/seahorse/issues/204, https://github.com/EspressoSystems/seahorse/issues/205

### CODE REVIEW

File: src/cli.rs [hunk 1]
```diff
@@ -626,10 +626,10 @@ fn init_commands<'a, C: CLI<'a>>() -> Vec<Command<'a, C>> {
                                 }
                             };
                             // Try to get a readable name for the asset.
-                            let asset = if txn.asset == AssetCode::native() {
+                            let asset = if *txn.asset() == AssetCode::native() {
```
Comment (jbearer) [inline src/cli.rs:629]: Idea for cleaning this sort of thing up a little: https://github.com/EspressoSystems/seahorse/issues/243

File: src/cli.rs [hunk 2]
```diff
@@ -626,10 +626,10 @@ fn init_commands<'a, C: CLI<'a>>() -> Vec<Command<'a, C>> {
...
                                 String::from("Native")
                             } else if let Some(asset) = keystore
-                                .asset(txn.asset)
+                                .asset(*txn.asset())
                                 .await
                             {
                                 if let Some(mint_info) = asset.mint_info() {
```
Comment (jbearer) [inline src/cli.rs:635]: Also unrelated to this PR, but: why are we not using `asset.name()` here? If it's what I think (that this code predates `Asset::name` and `Asset::description`) we should probably rework this a little bit. cc @shenkeyao

File: src/key_scan.rs [hunk 3]
```diff
@@ -372,11 +373,17 @@ pub fn receive_history_entry<L: Ledger>(
     };
 
     let txn_asset = last_record.asset_def.code;
-    TransactionHistoryEntry {
+    TransactionParams::<L> {
+        uid: Some(uid),
+        timeout: None,
+        status: TransactionStatus::Unknown,
```
Comment (jbearer) [inline src/key_scan.rs:379]: I think this should be `TransactionStatus::Retired` -- if we're receiving the transaction, it must have been finalized, right?

File: src/key_scan.rs [hunk 4]
```diff
@@ -372,11 +373,17 @@ pub fn receive_history_entry<L: Ledger>(
...
+        memos: Default::default(),
+        sig: None,
+        inputs: Default::default(),
+        outputs: Default::default(),
```
Comment (jbearer) [inline src/key_scan.rs:383]: I wonder if this should be `records`, since we actually have access to the record openings? Could be an eventual GUI feature, to click Learn More and see the output records of a transaction.

File: src/persistence.rs [hunk 5]
```diff
@@ -283,25 +271,25 @@ impl<'a, L: Ledger, Meta: Send + Serialize + DeserializeOwned> AtomicKeystoreSto
...
+    // }
+
+    // pub async fn transaction_history(
+    //     &mut self,
+    // ) -> Result<Vec<TransactionHistoryEntry<L>>, KeystoreError<L>> {
+    //     self.txn_history
+    //         .iter()
+    //         .map(|res| res.context(crate::PersistenceSnafu))
+    //         .collect()
+    // }
```
Comment (jbearer) [inline src/persistence.rs:292]: Delete

File: src/persistence.rs [hunk 6]
```diff
@@ -493,11 +473,12 @@ mod tests {
             key: KeyTree::random(&mut rng).0,
         };
         {
-            let (mut atomic_store, mut storage, mut assets) =
+            let (mut atomic_store, mut storage, mut assets, mut transactions) =
```
Comment (jbearer) [inline src/persistence.rs:476]: I don't like this pattern. We don't _need_ to create all the various resource types in order to create an `AtomicKeystoreStorage` and an associated `AtomicStore`. Doing so couples these tests to the behavior of other resources, and forces you to modify this seemingly unrelated test code whenever you add a new resource type.

I changed this in #231 already though, so feel free to leave as is and I'll address in that PR when I fix merge conflicts. Unless you can think of a reason why it should be this way, then just let me know.

File: src/transactions.rs [hunk 7]
```diff
@@ -16,17 +16,20 @@ use crate::{
...
+use serde::{Deserialize, Serialize};
 use std::collections::HashSet;
 use std::ops::{Deref, DerefMut};
 
 /// A Transaction<L>with its UID as the primary key.
-#[derive(Clone, Debug, Deserialize, Serialize)]
+#[derive(Clone, Debug, Derivative, Deserialize, Serialize)]
+#[serde(bound = "")]
+#[derivative(PartialEq(bound = "L: Ledger"))]
 pub struct Transaction<L: Ledger> {
```
Comment (jbearer) [inline src/transactions.rs:33]: Looking over the rest of this code, I find myself getting a little confused occasionally about which fields are set for received transactions, sent transactions, observed transactions....

Do you have any ideas about how we could clarify this? A good starting point would just be to include in the Rust doc for each field the conditions under which it is present. We might be able to do better, at least for some fields, by having sub-structures like `SentTransaction` for the sender-only fields and `ReceivedTransaction`, etc. Then in `Transaction`, instead of a bunch of related-by-convention fields, we could just have like `sender_info: Option<SentTransaction>`

File: src/transactions.rs [hunk 8]
```diff
@@ -39,7 +42,7 @@ pub struct Transaction<L: Ledger> {
     pending_uids: HashSet<u64>,
     /// A receiver memo for each output, except for burned outputs.
     memos: Vec<Option<ReceiverMemo>>,
-    sig: Signature,
+    sig: Option<Signature>,
```
Comment (jbearer) [inline src/transactions.rs:45]: Since `memos` and `sig` are related (they should always be both present or neither) does it make sense to do `memos: Option<(Vec<Option<ReceiverMemo>>, Signature)>`? Maybe even create a little struct like `SignedMemos { memos, sig }`

File: src/transactions.rs [hunk 9]
```diff
@@ -97,10 +100,13 @@ impl<L: Ledger> Transaction<L> {
     pub fn timeout(&self) -> Option<u64> {
         self.timeout
     }
+    pub fn hash(&self) -> &Option<TransactionHash<L>> {
```
Comment (jbearer) [inline src/transactions.rs:103]: Nit: it's probably a tiny bit more conventional from an API perspective to return `self.hash.as_ref() : Option<&TransactionHash<L>>`. See e.g. [HashMap::get](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.get) or [Vec::last](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.last).

File: src/transactions.rs [hunk 10]
```diff
@@ -133,22 +139,32 @@ impl<L: Ledger> Transaction<L> {
...
-impl<'a, L: Ledger + Serialize + DeserializeOwned> TransactionEditor<'a, L> {
+impl<'a, L: Ledger> TransactionEditor<'a, L> {
     /// Create a transaction editor.
     fn new(store: &'a mut Transactions<L>, transaction: Transaction<L>) -> Self {
         Self { transaction, store }
     }
 
+    pub fn transaction(&self) -> &Transaction<L> {
+        &self.transaction
+    }
```
Comment (jbearer) [inline src/transactions.rs:166]: I think this is redundant because of the `Deref` impl

File: src/transactions.rs [hunk 11]
```diff
@@ -184,16 +200,20 @@ impl<'a, L: Ledger + Serialize + DeserializeOwned> TransactionEditor<'a, L> {
     }
 
     /// Add the UIDs of memos we are waiting on to complete the transaction
-    pub fn add_pending_uids(mut self, uids: &[u64]) -> Self {
+    pub fn add_pending_uids(mut self, uids: Vec<u64>) -> Self {
```
Comment (jbearer) [inline src/transactions.rs:203]: Why this change? If we don't require ownership of the `Vec`, taking a slice is a bit better because it allows the function to be called in more situations without cloning.

File: src/transactions.rs [hunk 12]
```diff
@@ -205,37 +225,50 @@ impl<'a, L: Ledger + Serialize + DeserializeOwned> TransactionEditor<'a, L> {
...
+    pub fee_change: Option<RecordAmount>,
+    pub asset_change: Option<RecordAmount>,
+}
+
+impl<L: Ledger> PartialEq<Self> for TransactionParams<L> {
+    fn eq(&self, other: &Self) -> bool {
+        self.uid == other.uid
+            && self.kind == other.kind
+            && self.timeout == other.timeout
+            && self.time == other.time
```
Comment (jbearer) [inline src/transactions.rs:266]: This seems incomplete. Is it possible to use `derivative` to derive `PartialEq` so we don't have to worry about updating this whenever we add a field?

File: src/transactions.rs [hunk 13]
```diff
@@ -367,18 +400,31 @@ impl<L: Ledger + Serialize + DeserializeOwned> Transactions<L> {
...
-                    .clear_timeout()
-                    .save()?;
+                // Only return transactions that are not completed because other
+                // transactions did not actually expire.  We still remove all transactions
+                // from they index becuase we'll never lookup for this timeout again.
+                // Also update the status of the expired transactions
+                if editor.transaction.status() == TransactionStatus::Pending {
+                    removed.push(editor.transaction.clone());
+                    editor
+                        .set_status(TransactionStatus::Rejected)
```
Comment (jbearer) [inline src/transactions.rs:418]: We could actually have a `TimedOut` status now. Before we couldn't because querying the status on demand, we couldn't differentiate between a timed out transaction and an invalid one. Something to consider later on, maybe.

File: src/transactions.rs [hunk 14]
```diff
@@ -411,7 +457,7 @@ impl<L: Ledger + Serialize + DeserializeOwned> Transactions<L> {
         params: TransactionParams<L>,
     ) -> Result<TransactionEditor<'_, L>, KeystoreError<L>> {
         let txn = Transaction::<L> {
-            uid: params.uid,
+            uid: params.uid.unwrap(),
```
Comment (jbearer) [inline src/transactions.rs:460]: Not a huge fan of this `unwrap`. For required params, I think it would be better if we could take them as parameters of `create`, and save `TransactionParams` for optional stuff.

File: src/lib.rs [hunk 15]
```diff
@@ -645,7 +630,7 @@ impl<'a, L: 'static + Ledger> KeystoreState<'a, L> {
         model: &mut KeystoreModel<'a, L, impl KeystoreBackend<'a, L>, Meta>,
         receipt: &TransactionReceipt<L>,
     ) -> Result<TransactionStatus, KeystoreError<L>> {
-        match self.txn_state.transactions.status(&receipt.uid) {
+        match model.transactions.get(&receipt.uid)?.status() {
```
Comment (jbearer) [inline src/lib.rs:633]: Honestly, I would be in favor of getting rid of this entire mechanism, and making `type TransactionReceipt<L> = TransactionHash<L>`. This has never really worked properly for foreign transactions, and anyway it leaks privacy to call this for a transaction you expect to receive. For our own transactions, meanwhile, we now have the transaction status saved and updated, so this can all be replaced with `keystore.transactions().get(hash).status()`.

If you really want to track a foreign transaction by its hash, you can use a block explorer for that.

File: src/lib.rs [hunk 16]
```diff
@@ -852,22 +845,23 @@ impl<'a, L: 'static + Ledger> KeystoreState<'a, L> {
...
-
             LedgerEvent::Memos {
                 outputs,
                 transaction,
             } => {
                 let completed = self
                     .txn_state
-                    .transactions
-                    .received_memos(outputs.iter().map(|info| info.2));
+                    .received_memos(outputs.iter().map(|info| info.2), &mut model.transactions);
```
Comment (jbearer) [inline src/lib.rs:864]: Why does this go through `txn_state` at all now? The transaction _builder_ should operate only on the ledger state, not pending transactions, and indeed if I'm looking at the right thing, `TransactionState::received_memos` doesn't use `self`. It seems like this should be replaced with a direct invocation of the `Transactions` model, like you did above with `add_pending_uids`. (Sort of inlining some logic that will eventually make up the `TransactionsController`)

File: src/lib.rs [hunk 17]
```diff
@@ -1052,17 +1052,18 @@ impl<'a, L: 'static + Ledger> KeystoreState<'a, L> {
...
 
         if let Err(err) = model
-            .store(|mut t| async move {
-                t.store_transaction(history).await?;
-                Ok(t)
-            })
-            .await
+            .transactions
+            .create(history)
+            .unwrap()
```
Comment (jbearer) [inline src/lib.rs:1064]: Definitely should not be unwrapping here. This will panic if there's a temporary error in persistence. Just fall through to the println for now.

(This is another thing that's improved in #231, which allows this function to return an error)

File: src/lib.rs [hunk 18]
```diff
@@ -1129,39 +1130,44 @@ impl<'a, L: 'static + Ledger> KeystoreState<'a, L> {
...
         &mut self,
-        txn: &Transaction<L>,
+        model: &mut KeystoreModel<'a, L, impl KeystoreBackend<'a, L>, Meta>,
+        txn: &reef::Transaction<L>,
         res: Option<CommittedTxn<'t>>,
-    ) -> Option<PendingTransaction<L>> {
-        let pending = self.txn_state.clear_pending_transaction(txn, &res);
+    ) -> Option<Transaction<L>> {
+        let pending = self
+            .txn_state
```
Comment (jbearer) [inline src/lib.rs:1140]: This one appears to go through `txn_state` because it has an interaction with records. Seems fine for now. Exactly the kind of thing that we should hope to clean up with the controller layer.

**Review by jbearer — COMMENTED** [review]
Looking really good. I left a bunch of suggestions and questions but nothing major. Love to see that negative line count!

File: src/key_scan.rs (hunk 4 above)
Comment (bfish713) [reply src/key_scan.rs:383]: Good point, I left if empty because it `TransactionHistoryEntry` didn't have it but no reason not to add it.

File: src/lib.rs (hunk 15 above)
Comment (bfish713) [reply src/lib.rs:633]: Yeah I will get rid of this

File: src/persistence.rs (hunk 6 above)
Comment (bfish713) [reply src/persistence.rs:476]: Don't mind this changing.  I'll leave it for your PR though.

File: src/transactions.rs (hunk 7 above)
Comment (bfish713) [reply src/transactions.rs:33]: The only way I can think of is to have optional send and receive structs and optional fields for them in `Transactions`.

File: src/transactions.rs (hunk 11 above)
Comment (bfish713) [reply src/transactions.rs:203]: I will take another look at this.  I was having some trouble compiling this when the param was a slice.  But now I can't remember the exact issue.  Revert this change or post the issue if I hit it and can't solve it.

File: src/transactions.rs (hunk 12 above)
Comment (bfish713) [reply src/transactions.rs:266]: I will try this!

File: src/transactions.rs (hunk 14 above)
Comment (bfish713) [reply src/transactions.rs:460]: Yeah I want to change this.  I was thinking I probably need to create some other type to store the transaction info as we build the transaction, then do as you said have create take the optional params as `TransactionParams` and the required params directly.

File: src/txn_builder.rs [hunk 19]
```diff
@@ -1109,31 +710,32 @@ impl<L: Ledger> TransactionState<L> {
...
+        txn: &reef::Transaction<L>,
         res: &Option<CommittedTxn<'t>>,
-    ) -> Option<PendingTransaction<L>> {
+    ) -> Option<Transaction<L>> {
         let now = self.validator.now();
 
-        // Remove the transaction from pending transaction data structures.
         let txn_hash = txn.hash();
-        let pending = self.transactions.remove_pending(&txn_hash);
+        let pending = transactions.with_hash(&txn_hash).ok();
```
[the remark this answers is not in the document]
Comment (bfish713) [reply src/txn_builder.rs:738]: 1. Yes that's the case and we would keep it around
2. Yes this should be taken care of by revert.  My understanding/intention of how it works with my changes if the transaction fails to be submited we call `transactions.revert` which should remove the transaction from memory and from the `txn_by_hash` index.  I actually think the bug is that we call `revert` before we call this function meaning we won't unhold the records used for the transaction.  That's an easy fix.

File: src/transactions.rs (hunk 7 above)
Comment (jbearer) [reply src/transactions.rs:33]: Yeah. Should we table this for future work?

File: src/transactions.rs [hunk 20]
```diff
@@ -41,8 +48,7 @@ pub struct Transaction<L: Ledger> {
     /// The uids of the outputs of this transaction for which memos have not yet been posted.
     pending_uids: HashSet<u64>,
     /// A receiver memo for each output, except for burned outputs.
```
Comment (jbearer) [inline src/transactions.rs:50]: I think this comment should be moved up to `SignedMemos::memos`, since it explains the somewhat surprising type `Vec<Option<ReceiverMemo>>`

File: src/transactions.rs [hunk 21]
```diff
@@ -454,16 +442,17 @@ impl<L: Ledger> Transactions<L> {
...
+        uid: TransactionUID<L>,
+        hash: TransactionHash<L>,
         params: TransactionParams<L>,
     ) -> Result<TransactionEditor<'_, L>, KeystoreError<L>> {
         let txn = Transaction::<L> {
-            uid: params.uid.unwrap(),
+            uid,
             timeout: params.timeout,
-            hash: None,
+            hash: Some(hash),
```
Comment (jbearer) [inline src/transactions.rs:452]: Does `hash` still need to be an `Option`, if we get it as a parameter for `create`? (BTW when we do #238, `hash` and `uid` will always be the same, so we can remove one of them and this will get a bit simpler)

File: src/lib.rs [hunk 22]
```diff
@@ -1622,21 +1610,28 @@ impl<'a, L: 'static + Ledger> KeystoreState<'a, L> {
                 let t = self
                     .txn_state
                     .add_pending_transaction(&mut model.transactions, &txn, info);
-                if let Err(err) = model.transactions.commit() {
+                if let Err(err) = model
```
Comment (jbearer) [inline src/lib.rs:1613]: Now `transactions.commit()` is not getting called?

File: src/lib.rs [hunk 23]
```diff
@@ -2727,53 +2722,71 @@ impl<
...
+            // Transaction uids are unique only to a given keystore, so if we're trying to track
+            // somebody else's transaction, the best we can do is wait for one of its nullifiers
+            // to be published on the ledger.
+            pending_foreign_txns
+                .entry(receipt.fee_nullifier)
+                .or_insert_with(Vec::new)
+                .push(sender);
+            drop(guard);
+            receiver.await.map_err(|_| KeystoreError::<L>::Cancelled {})
+        }
```
Comment (jbearer) [inline src/lib.rs:2761]: I thought we were getting rid of this behavior. If we do, we can also get rid of `TransactionReceipt` (because we would only ever use `receipt.uid`) and some of the complicated logic with `pending_foreign_txns` in the event handling thread. Any reason to keep this?

File: src/lib.rs [hunk 24]
```diff
@@ -2727,53 +2722,71 @@ impl<
...
     ) -> Result<TransactionStatus, KeystoreError<L>> {
         let mut guard = self.mutex.lock().await;
         let KeystoreSharedState {
-            state,
-            model,
+            state: _,
+            model: _,
             txn_subscribers,
-            pending_foreign_txns,
+            pending_foreign_txns: _,
```
Comment (jbearer) [inline src/lib.rs:2775]: Probably just get rid of the `field: _` if they are unused. The `..` below already means "ignoring all other fields"

File: src/transactions.rs (hunk 21 above)
Comment (bfish713) [reply src/transactions.rs:452]: The only reason I left the hash in was because I thought for some other ledger types Hash and UID could be different.  That said now that I am looking at it is seems like there isn't any way to decouple them.  I think removing UID makes sense since it is derived from the hash.

File: src/lib.rs (hunk 23 above)
Comment (bfish713) [reply src/lib.rs:2761]: Only reason to keep receipt was because the CLI uses receipt in it's output but I guess if we have the hash we can do whatever we need.

This function we can get rid of.  I don't really know why I kept it

Comment (jbearer) [reply src/lib.rs:2761]: Nice! This is one of the parts I always hated most. And yeah, the CLI can just output the transaction hash (which makes us a little bit more consistent with how most Ethereum stuff works, anyway)

File: src/lib.rs (hunk 22 above)
Comment (bfish713) [reply src/lib.rs:1613]: it's called in `model.store()`

Comment (jbearer) [reply src/lib.rs:1613]: Ah, good

File: src/transactions.rs (hunk 21 above)
Comment (bfish713) [reply src/transactions.rs]: Decided to keep UID in the transaction because it has some stuff already built in and will be easier to extend if we need to because it's defined in this repo.

File: src/lib.rs [hunk 25]
```diff
@@ -2701,52 +2722,38 @@ impl<
...
-            state,
+            state: _,
             model,
             txn_subscribers,
-            pending_foreign_txns,
             ..
         } = &mut *guard;
-
-        let status = state.transaction_status(model, receipt).await?;
+        let status = model.transactions.get(uid)?.status();
```
Comment (bfish713) [inline src/lib.rs:2746]: @jbearer The cause of the deadlock was actually here not with `AtomicStore::commit_version` for the most recent test timeouts.  I had changed this to call `self.transaction_status()` which locks the shared state again

**Review by jbearer — APPROVED** [review]
I think there should be some dead code now -- the stuff dealing with `pending_foreign_txns` and `foreign_txns_awaiting_memos` in `new_impl`. Otherwise looks good
